### PR TITLE
[Close #335] Show asset compiling on error

### LINKED ISSUE #335: Show asset currently compiling when error
This error was generated while compiling an asset, which asset? No one knows! We should consider catching all errors, extending the error message and re-raising.

```
       rake aborted!
       NoMethodError: undefined method `protocol' for nil:NilClass
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/actionview-4.2.6/lib/action_view/helpers/asset_url_helper.rb:227:in `compute_asset_host'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/actionview-4.2.6/lib/action_view/helpers/asset_url_helper.rb:143:in `asset_path'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/actionview-4.2.6/lib/action_view/helpers/asset_url_helper.rb:162:in `asset_url'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/exception_handler-0.5.1/app/assets/stylesheets/exception_handler/error.css.erb:26:in `_evaluate_template'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/erb_processor.rb:26:in `call'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/erb_processor.rb:13:in `call'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/processor_utils.rb:75:in `call_processor'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/processor_utils.rb:57:in `block in call_processors'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/processor_utils.rb:56:in `reverse_each'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/processor_utils.rb:56:in `call_processors'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/loader.rb:134:in `load_from_unloaded'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/loader.rb:60:in `block in load'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/loader.rb:44:in `load'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/cached_environment.rb:20:in `block in initialize'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/cached_environment.rb:47:in `load'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/bundle.rb:23:in `block in call'
       /tmp/build_f9e13c24ea425636297c8723889e488a/vendor/bundle/ruby/2.3.0/gems/sprockets-3.6.2/lib/sprockets/utils.rb:196:in `dfs'
```

### PR DESCRIPTION
**Pull request by schneems**
When there's an error compiling, the least we can do is mention what asset we're currently compiling.

### CODE REVIEW

**Comment by matthewd** [comment]
This pretty horribly mangles the original exception.

Isn't the erroneous asset [source] already in the exception backtrace?

(I think this would belong closer to `load` anyway, maybe?)

**Comment by schneems** [comment]
> Isn't the erroneous asset [source] already in the exception backtrace?

No. In #335 the exception comes from ERB and is a pure Ruby exception, it knows nothing about sprockets there is no source in it. For cases where there is an exception coming somewhere other than sprockets there is basically no debug info. Even in cases where we are raising a sprockets exception such as a `Sprockets::LoadError` there is still value in providing the original top level asset that we are trying to load. For example the load error might be on `bar.js` but there is no indication of what was trying to load it. This exception mangling will tell you that it was `foo.js` that was trying to load it which is just as important as what couldn't be loaded.

Exceptions can come from processors, transformers and from Sprockets itself. There may be cases where we have duplicate information i.e. we are trying to load `bar.js` and a load error for `bar.js` gets thrown, however in that situation having duplicate information won't hurt.

> This pretty horribly mangles the original exception.

Here's other alternatives

0) Do nothing: This failure scenario where something raised an exception outside of sprockets deserves a better debugging scenario. An easy piece of information is where you were in the asset compilation story when this happened.

1) Warn instead of mangling the exception: I don't like this because what if a top level library is catching the exception...then it wouldn't catch the warning.

2) Raise a custom error that wraps the original error: I don't think this helps at all.

3) Provide an API for disabling the error mangling: this is do-able, I don't know how used/useful it would be.

4) I'm open to other suggestions.

> (I think this would belong closer to load anyway, maybe?)

The problem with that is load can be deeply deeply nested. For debugging I want the most top level asset possible. I'm not opposed to explicitly adding other information but before I do I want to fully understand the failure scenarios.

File: test/fixtures/default/error_in_erb.js.erb [hunk 1]
```diff
@@ -0,0 +1 @@
+var foo = <%= nil.no_method %>;
```
Comment (jvanbaarsen) [inline test/fixtures/default/error_in_erb.js.erb]: Do we need a newline here?

**Comment by schneems** [comment]
@matthewd can you take a look at this change. I ended up wrapping all exceptions in a new exception `Sprockets::LoadError`. I'm a bit worried that we're going to break any code that was rescuing a specific exception such as `Sass::SyntaxError`, but I don't think we can avoid that.

**Comment by matthewd** [comment]
Are you sure we don't *want* a deeply nested set of exceptions, if that's the nature of the problem?

Knowing that we were trying to compile `a.js` when we failed to find `err.js` is interesting.. but it seems even more useful to know that `a.js` required `b.js` required `c.js` required `err.js`. Otherwise, we're still not telling you which line actually pulled in the bad filename -- assume `a.js` is in fact named `application.js`, and... :confused:

(this reminds me of [a past time](https://github.com/matthewd/stackdeck))

**Comment by schneems** [comment]
There is already functionality to adding a line to the backtrace in the directive processor https://github.com/rails/sprockets/blob/c9ab1b45d560a5527caba1e0815ef2e6953fce51/lib/sprockets/directive_processor.rb#L183 however there's two issues 

1) Nested backtraces show up in reverse order. 

2) When the directive processor calls `process_require_directive` the file is only required and not actually loaded. The loading (where the ERB exception happens) is coming from the bundler processor. Which is why the full backtrace isn't being showed now. I played around and added a `load` call right after require and all tests pass except for the circular-require test. I'll play with it some more.

Putting the info in the backtrace and not messing around or mangling the error seems like a good way forward.

**Comment by matthewd** [comment]
> Nested backtraces show up in reverse order.

Reverse as in the inner-most is at the top, like a normal backtrace? Or reversed from normal backtrace order? :confused:

> When the directive processor calls process_require_directive the file is only required and not actually loaded. The loading (where the ERB exception happens) is coming from the bundler processor. Which is why the full backtrace isn't being showed now. I played around and added a load call right after require and all tests pass except for the circular-require test. I'll play with it some more.

Perhaps `process_require_directive`, while noting that a given file is required, can also record the file/line it was required *by*. Then load can insert the suitably-blaming backtrace line when it hits the exception, even though the blamed file isn't *actually* being processed at that time.

**Comment by schneems** [comment]
> Reverse as in the inner-most is at the top, like a normal backtrace? Or reversed from normal backtrace order? 😕

The order it produces is not correct, so we would need to capture all the line information then reverse it before putting it on top of the backtrace. It's do-able i think.

> Perhaps process_require_directive, while noting that a given file is required, can also record the file/line it was required by. Then load can insert the suitably-blaming backtrace line when it hits the exception, even though the blamed file isn't actually being processed at that time.

I got stuck so I spent some time tracing exactly how sprockets loads an asset. I'll polish and publish eventually, here's the raw notes if you're interested https://draftin.com/documents/970030?token=oSLGZTgJDghg1i13RfPtrhDsw7irXHLvRhkNxFcRv2_1ech7wEDKs3Auz2Qc_c5RA7kkdBY7gPSpDg4qpj8n6Ys

I think we can maybe use the same thing that the `Bundle` processor uses to prevent circular requires `Utils.dfs` inside of the `DirectiveProcessor.resolve`, I spent too long writing the doc, and didn't have time to test it. That's my next approach though. Just wanted to report on my progress, or lack thereof.

**Pull request closed without merging**